### DISCUSSION TITLE
Please advise: Copy tables from Oracle to sqlite

### DISCUSSION THREAD

**Discussion by nklamann** [sqlalchemy/sqlalchemy#9558] [Usage Questions] [answered]
Hallo all,
for monitoring purposes I want to download some Views of oracles huge Data dictionary and store them in a local sqlite database.
I implemented such a thing already in an oracle database but find this approach rather heavy-handed . 

I want to take **most** of the columns (but not all) and amend them with some metadata like timestamp.

I don't want to hard code the dozens of column names and plan to use the following process for each Oracle Data dictionary view in which I am interested:

1. Get Column list of the View
2. Create a copy of this metastructure
3. delete some columns
4. add some columns
5. write the table structure to the target database

On the same basis I would then generate the SQL to load the data.

This is my first project with SQLAlchemy and I like the idea to 'automate' the SQL but  not to hide it. 

I am pretty sure that some people did things like that already, maybe with other databases but the same principle, if it is valid at all.

Any pointers ?

Thanks a lot !

Norbert

**Comment by zzzeek** [comment, marked as answer]
sure, you will want to build up Table objects that match the schema you want to manipulate on SQLite.   When you build those Table objects and want to fill them up with Column objects, you can get those lists of columns / datatypes using the inspector:

https://docs.sqlalchemy.org/en/20/core/reflection.html#sqlalchemy.engine.reflection.Inspector.get_columns

see https://docs.sqlalchemy.org/en/20/core/reflection.html#fine-grained-reflection-with-inspector for how to get data about database objects which you can use to build up Table objects and other things.

**Reply by nklamann** [reply]
Many thanks, but I think I have to adapt the column elements from Oracle to sqlite in any form.

When I call the Inspector to get the column list from the oracle side (like so `insp.get_columns("meta_object_columns", schema="BETRIEB") i get a list with dictionaries like the following
```python
{'name': 'object_name',
 'type': VARCHAR(length=100),
 'nullable': False,
 'default': None,
 'comment': None}
```
What is the best way to create SQLAlchemy columns from this data ?

Sorry, if this seems obvious but I did not find an answer.

EDIT:

I created this gist https://gist.github.com/nklamann/6f37d8a02f94281ebd80e4e6b6ea4cf5

**Reply by zzzeek** [reply]
you would write a loop similar to:


```py

real_columns = []
for col in inspector.get_columns("table"):
    real_column = Column(col['name'], col['type'].as_generic(), nullable=col['nullable'], server_default=col['default'])
    real_columns.append(real_column)

new_table = Table("table", metadata, *real_columns)
```

please note I'm giving you the low level "inspector" API because you said you want to pick and choose which columns you want, this gives you the most flexibility in this regard, as well as to be able to convert any Oracle-specific elements to SQLite, or to skip them, such as the "server_default" above which may not be cross-compatible, if your tables are using those or not.

if you want a quick "copy the table" you can use regular reflection to get a `Table` object all at once: https://docs.sqlalchemy.org/en/20/core/reflection.html

**Reply by nklamann** [reply]
Thanks, that helps a lot.
